torch/utils/data/datapipes/iter/combinatorics.py

```diff
@@ -113,21 +115,46 @@ def set_shuffle(self, shuffle=True):
         return self
 
     def __iter__(self) -> Iterator[T_co]:
+        self.reset()
         if not self._enabled:
             for x in self.datapipe:
                 yield x
         else:
-            buffer: List[T_co] = []
             for x in self.datapipe:
-                if len(buffer) == self.buffer_size:
-                    yield ShufflerIterDataPipe.buffer_replace(buffer, x)
+                if len(self.buffer) == self.buffer_size:
+                    yield ShufflerIterDataPipe.buffer_replace(self.buffer, x)
                 else:
-                    buffer.append(x)
-            random.shuffle(buffer)
-            while buffer:
-                yield buffer.pop()
+                    self.buffer.append(x)
+            random.shuffle(self.buffer)
+            while self.buffer:
+                yield self.buffer.pop()
 
     def __len__(self) -> int:
         if isinstance(self.datapipe, Sized):
             return len(self.datapipe)
         raise TypeError("{} instance doesn't have valid length".format(type(self).__name__))
+
+    def reset(self) -> None:
+        self.buffer = []
+
+    def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+        state = (
+            self.datapipe,
+            self.buffer_size,
+            self._enabled
+        )
+        return state
+
+    def __setstate__(self, state):
+        (
+            self.datapipe,
+            self.buffer_size,
+            self._enabled
+        ) = state
+        self.reset()
```

Comment:
Currently, the logic of `__getstate__` and `__setstate__`  isn't saving the content of the buffer. I am planning to add `save_snapshot` to do that, but open to do that within `__getstate__` instead.

Therefore, the reset acts as "re-initialization" here, and basically just sets the buffer back to empty.

Reply:
`buffer` is never created during "re-initialization". Can I know the reason behind splitting the functionality into `__getstate__` and `save_snapshot`?

Reply:
> In `DataLoader`, IIUC serialization typically is used by multiprocess and that happens in the beginning, such that the buffers are.

This is true. If we assume the buffer is empty, we don't necessarily need to run `reset` before serialization. Then, these two APIs are going to return the same thing.
If the `buffer` is not empty during serialization because of MP, I believe it should be users' intention. We should allow them to do so.

See the example from TorchVision:
https://github.com/pytorch/vision/blob/c67a58396d894c48897b96fce4eae0bcb3fd2755/torchvision/prototype/datasets/_builtin/cub200.py#L183-L185
They use `demux` to split image and meta data.
https://github.com/pytorch/vision/blob/c67a58396d894c48897b96fce4eae0bcb3fd2755/torchvision/prototype/datasets/_builtin/cub200.py#L188-L190
Then, they use the `dict` to iterate over the meta data pipeline to create a dictionary of image_id to path. At this point, the buffer inside `demux` should be full of data. If they put this data pipeline into MP, you will clean up buffer for them. This is non-intuitive to me.

Reply:
Thanks for the helpful TorchVision example. The issue discussed here is pre-existing as you can see from `master` branch's demux code snippet below. In the past, we have avoided serializing the buffer (`dp2 = deserialize(serialize(dp1))` returns a `dp2` that has an empty buffer). This is true for all DataPipes with buffers, but we should definitely change it for the better if it doesn't cause other issues.
https://github.com/pytorch/pytorch/blob/9cb4e1cbf9e3506dabdb396368f2156ce3fbe6a2/torch/utils/data/datapipes/iter/combining.py#L353-L378

I think the best course of action is to add a separate PR in the stack that change DataPipes from not serializing the buffers to serializing them (often removing custom `__getstate__` and `__setstate__` should do the trick), and test those changes properly.

I think with the introduction with the "serialization wrapper DataPipe" from the other PR, it should work without issue. We can also provide guidance to users to create a custom "serialization wrapper" for DataPipes if the existing one doesn't mean their needs.

cc: @VitalyFedyunin

Reply:
Actually I realize saving the buffer for `fork` and `demux` is a bad idea unless the snapshotting is complete, because without snapshotting, after the serialization/deserialization process, `main_datapipe` will read from the beginning again.

Let's say:
```python
main_dp = IterableWrapper(range(10))
cdp1, cdp2 = main_dp.fork(num_instances=2)
# cdp1, cdp2 = main_dp.demux(num_instances=2, classifier_fn=class_fn)
it1, it2 = iter(cdp1), iter(cdp2)
print(next(it1))
# Now we serialize and deserialize
new_cdp1 = pickle.loads(pickle.dumps(cdp1))
new_cdp2 = pickle.loads(pickle.dumps(cdp2))
# `main_dp` will restart from the beginning, because snapshotting isn't done yet, but the buffer will still have [0]
print(list(new_cdp1))  # We have some pointer logic that avoids element in the buffer if the child is the fastest
# [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
print(list(new_cdp2))
# [0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

We will have to address this with the full snapshotting.